### PR TITLE
Enable saving results only at the end of the simulation

### PR DESCRIPTION
**Pull request by nick-harder**
-adds a function to store all results at the end
-fixes cvs_path creating an empty folder when not defined

### CODE REVIEW

File: assume/common/outputs.py [hunk 1]
```diff
@@ -40,18 +40,18 @@ def __init__(
         start: datetime,
         end: datetime,
         db_engine=None,
-        export_csv_path: str = "",
-        save_frequency_hours: int = 24,
+        export_csv_path: str = None,
+        save_frequency_hours: int = None,
```
Comment (maurerle) [inline assume/common/outputs.py:44]: export_csv_path = "" did the job, why the change..?

File: assume/common/outputs.py [hunk 2]
```diff
@@ -40,18 +40,18 @@ def __init__(
...
         learning_mode: bool = False,
     ):
         super().__init__()
 
         # store needed date
         self.simulation_id = simulation_id
-        self.save_frequency_hours = save_frequency_hours
+        self.save_frequency_hours = save_frequency_hours or (end - start).days * 24
```
Comment (maurerle) [inline assume/common/outputs.py:51]: When would you want to disable saving results in given intervals?
Stating "same_process=False" should fix consuming performance, by using a separate process

The remaining data will always be written in `on_stop` of the outputs role

File: assume/common/outputs.py [hunk 3]
```diff
@@ -40,18 +40,18 @@ def __init__(
...
 
         # make directory if not already present
-        self.export_csv_path = export_csv_path
+        self.export_csv_path = export_csv_path or 0
         if self.export_csv_path:
             self.p = Path(self.export_csv_path, simulation_id)
             shutil.rmtree(self.p, ignore_errors=True)
```
Comment (maurerle) [inline assume/common/outputs.py:57]: here we are removing the folder and recreating it properly

File: assume/common/outputs.py [hunk 4]
```diff
@@ -40,18 +40,18 @@ def __init__(
...
         super().__init__()
 
         # store needed date
         self.simulation_id = simulation_id
-        self.save_frequency_hours = save_frequency_hours
+        self.save_frequency_hours = save_frequency_hours or (end - start).days * 24
 
         # make directory if not already present
-        self.export_csv_path = export_csv_path
+        self.export_csv_path = export_csv_path or 0
```
Comment (maurerle) [inline assume/common/outputs.py:54]:
```
export_csv_path = ""
if export_csv_path:
  print("hi")
```

does the same, the `or 0` does not add anything..?

File: examples/examples.py [hunk 5]
```diff
@@ -7,7 +7,8 @@
 log = logging.getLogger(__name__)
 
 csv_path = "./examples/outputs"
-os.makedirs(csv_path, exist_ok=True)
+if csv_path:
+    os.makedirs(csv_path, exist_ok=True)
```
Comment (maurerle) [inline examples/examples.py:11]: csv_path is always defined above..
you should just remove the makedirs line entirely, as this is taken care of in outputs.py in line 58

**Review by maurerle — CHANGES_REQUESTED** [review]

File: assume/common/outputs.py (hunk 1 above)
Comment (nick-harder) [reply assume/common/outputs.py:44]: for me it always creates an empty folder in the root without a name and writes the files over there. So it doesn't work

File: assume/common/outputs.py (hunk 2 above)
Comment (nick-harder) [reply assume/common/outputs.py:51]: as of right now this saving each day consumes a big chunk of time as multithread is not working. We should have this until second thread works. Also, sometime you really don't need all that data every day of the run...

File: assume/common/outputs.py (hunk 3 above)
Comment (nick-harder) [reply assume/common/outputs.py:57]: as I said before, on my system defining export_csv_path as "" makes a no name folder. This fixes it

File: examples/examples.py (hunk 5 above)
Comment (nick-harder) [reply examples/examples.py:11]: cool

File: assume/common/outputs.py (hunk 3 above)
Comment (nick-harder) [reply assume/common/outputs.py:57]: so the problem was actually because of the mkdir in the main examples.py file and not here. I have now removed it from the main file and it seems to work fine

File: assume/common/outputs.py (hunk 2 above)
Comment (nick-harder) [reply assume/common/outputs.py:51]: I believe this is also not because of the db, but the csv files, those require a lot of time

**Comment by nick-harder** [comment]
@maurerle soll es durch?

File: assume/common/outputs.py (hunk 1 above)
Comment (maurerle) [reply assume/common/outputs.py:44]: I don't think that we should set this to None..

**Comment by maurerle** [comment]
Ich wollte es noch nachstellen, ob das = None tatsächlich notwendig ist..

**Comment by nick-harder** [comment]
ach ja stimmt, jetzt ist 24 wieder default

**Review by maurerle — APPROVED** [review]
